python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,1034 @@
+        quanta, or other entities that use a (autoincrement, origin) compound
+        primary key.
+    cs : `ConnectionStruct`
+        A struct containing connection and transaction state for the database.
+        May be shared between `Database` instances.
+    namespace : `str`, optional
+        Name of the schema or namespace this instance is associated with.
+        This is passed as the ``schema`` argument when constructing a
+        `sqlalchemy.schema.MetaData` instance.  We use ``namespace`` instead to
+        avoid confusion between "schema means namespace" and "schema means
+        table definitions".
+
+    Notes
+    -----
+    `Database` requires all write operations to go through its special named
+    methods.  Our write patterns are sufficiently simple that we don't really
+    need the full flexibility of SQL insert/update/delete syntax, and we need
+    non-standard (but common) functionality in these operations sufficiently
+    often that it seems worthwhile to provide our own generic API.
+
+    In contrast, `Database.query` allows arbitrary ``SELECT`` queries (via
+    their SQLAlchemy representation) to be run, as we expect these to require
+    significantly more sophistication while still being limited to standard
+    SQL.
+
+    `Database` itself has several underscore-prefixed attributes:
+
+     - ``_cs``: SQLAlchemy objects representing the connection and transaction
+        state.
+     - ``_metadata``: the `sqlalchemy.schema.MetaData` object representing
+        the tables and other schema entities.
+
+    These are considered protected (derived classes may access them, but other
+    code should not), and read-only, aside from executing SQL via
+    ``_cs.connection``.
+    """
+
+    def __init__(self, *, origin: int, cs: ConnectionStruct, namespace: Optional[str] = None):
+        self.origin = origin
+        self.namespace = namespace
+        self._cs = cs
+        self._metadata = None
+
+    @classmethod
+    def fromUri(cls, uri: str, *, origin: int, writeable: bool = True) -> Database:
+        """Construct a database from a SQLAlchemy URI.
+
+        Parameters
+        ----------
+        uri : `str`
+            A SQLAlchemy URI connection string.
+        origin : `int`
+            An integer ID that should be used as the default for any datasets,
+            quanta, or other entities that use a (autoincrement, origin)
+            compound primary key.
+        writeable : `bool`, optional
+            If `True`, allow write operations on the database, including
+            ``CREATE TABLE``.
+
+        Returns
+        -------
+        db : `Database`
+            A new `Database` instance.
+        """
+        return cls.fromConnectionStruct(cls.connect(uri, writeable=writeable), origin=origin,
+                                        writeable=writeable)
+
+    @classmethod
+    @abstractmethod
+    def connect(cls, uri: str, *, writeable: bool = True) -> ConnectionStruct:
```

Comment:
You guys move too fast, while I'm writing a review 5 new commits are made :). 

It used to be that one would instantiate a Registry from a RegistryConfig. It used to be that it was the job of RegistryConfig [connectionString](https://github.com/lsst/daf_butler/blob/master/python/lsst/daf/butler/core/registryConfig.py#L68) property to resolve the full connection from the YAML file, supplementing usernames, ports, password etc. from the `~/.lsst/db-auth.yaml` file as, if at all, necessary based on the db string provided in `butler.yaml`. 

Would it be too much to ask that `connect` optionally accepts a direct URI or an `RegistryConfig` instance? That way one can still depend on the `connectionString` property to resolve, or not, users and passwords. 

I understand, from the [architecture design doc](https://confluence.lsstcorp.org/display/DM/Architectural+Prototype+for+the+New+Gen3+Registry), that `Registry` has to go but it wasn't clear to me if that's the case for `RegistryConfig` also.

Reply:
Heh, you're moving too fast, commenting on my commits while I'm still amending them :smiley: .

Anyhow, I was mostly planning to tackle this in a future ticket (when I make `Registry` start to use `Database`), but I am certainly aiming to keep the high-level behavior the same.  I think I'd lean towards leaving these APIs as-is, but use `RegistryConfig` to generate the final connection string to pass to them.  It'd also be responsible for inferring the `Database` subclass (instead of the `Registry` subclass) from what's in the config.  Does that sound like it would work?

Reply:
That sounds perfect, in fact that sounds to work out exactly like it is right now. The correct registry class is inferred from the `RegistryConfig.getRegistryClass` based on  the mapping provided in `config/registry.yaml`(for future reference).